src/main/java/org/openjdk/jextract/impl/IncludeFilter.java

```diff
@@ -76,43 +69,35 @@ public Void visitScoped(Declaration.Scoped d, Declaration parent) {
         if (isStructKind) {
             String name = d.name();
             if (!name.isEmpty() && !includeHelper.isIncluded(d)) {
-                return null;
+                //skip
+                d.addAttribute(new Skip());
             }
         }
 
-        List<Declaration> oldDecls = decls;
-        this.decls = new ArrayList<>();
-        try {
-            d.members().forEach(fieldTree -> fieldTree.accept(this, d));
-        } finally {
-            var scoped = createScoped(d, decls);
-            this.decls = oldDecls;
-            decls.add(scoped);
-        }
+        d.members().forEach(fieldTree -> fieldTree.accept(this, d));
         return null;
     }
 
     @Override
     public Void visitTypedef(Declaration.Typedef tree, Declaration parent) {
         if (!includeHelper.isIncluded(tree)) {
-            return null;
+            //skip
+            tree.addAttribute(new Skip());
         }
-        decls.add(tree);
         return null;
     }
 
     @Override
     public Void visitVariable(Declaration.Variable tree, Declaration parent) {
         if (parent == null && !includeHelper.isIncluded(tree)) {
-            return null;
+            //skip
+            tree.addAttribute(new Skip());
```

Comment:
Maybe we could use a singleton instance?

Reply:
Yeah...